package/MDAnalysis/analysis/contacts.py

```diff
@@ -437,8 +441,26 @@ def __init__(self, u, select, refgroup, method="hard_cut", radius=4.5,
             self.fraction_contacts = method
 
         self.select = select
-        self.grA = u.select_atoms(select[0])
-        self.grB = u.select_atoms(select[1])
+
+        if isinstance(select[0], str):
+            self.grA = u.select_atoms(select[0])
+        elif isinstance(select[0], AtomGroup):
+            if isinstance(select[0], UpdatingAtomGroup):
+                raise TypeError(self.select_error_message)
+            else:
+                self.grA = select[0]
+        else:
+            raise TypeError(self.select_error_message)
+
+        if isinstance(select[1], str):
+            self.grB = u.select_atoms(select[1])
+        elif isinstance(select[1], AtomGroup):
+            if isinstance(select[1], UpdatingAtomGroup):
+                raise TypeError(self.select_error_message)
+            self.grB = select[1]
+        else:
+            raise TypeError(self.select_error_message)
```

Comment:
There's a lot of repeated code here, let's see if we can condense it a little bit:

How about something like:

```python
@staticmethod
def _get_atomgroup(u, entry):
    if isinstance(entry, str):
        return u.select_atoms(entry)
    elif:
        if isinstance(entry, AtomGroup):
            if isinstance(entry, UpdatingAtomGroup):
                errmsg = "UpdatingAtomGroups are not supported"
                raise TypeError(errmsg)
            else:
                return entry
    else:
        raise TypeError(f"unrecognised selection type {entry}")

# ... and now back in __init__

self.grA, self.grB = [_get_atomgroup(var) for var in select]
```

This reduces the number of lines and also means you can directly test the staticmethod without having to go via class creation.

Reply:
Done!